hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithPersistedDiskCache.java

```diff
@@ -0,0 +1,83 @@
+package se.liu.ida.hefquin.engine.federation.access.impl;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CacheEntry;
+import se.liu.ida.hefquin.base.datastructures.impl.cache.CachePolicies;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.access.CardinalityResponse;
+import se.liu.ida.hefquin.engine.federation.access.DataRetrievalResponse;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessException;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+
+/**
+ * A FederationAccessManager implementation that incorporates persistent disk
+ * caching of SPARQL cardinality requests.
+ * 
+ * TODO: The implementation uses a simple serialization/deserialization
+ * strategy, where the file is stored to disk by writing the full map to disk
+ * for each update. This approach is not optimized for this task but is simply
+ * intended as a proof of concept. A future implementation should support
+ * standard cache configuration policys, such as time-based eviction (time to
+ * live), but should leverage an optimized persistence strategy, preferably
+ * leveraging a libary.
+ * 
+ * Note: Most of the classes/interfaces involved (e.g., DataRetrievalResponse,
+ * CardinalityResponse etc.) do not support serialization.
+ */
+public class FederationAccessManagerWithPersistedDiskCache extends FederationAccessManagerWithCache
+{
+	protected CardinalityCacheMap<CardinalityCacheKey> cardinalityCache;
+
+	public FederationAccessManagerWithPersistedDiskCache( final FederationAccessManager fedAccMan,
+	                                                      final int cacheCapacity,
+	                                                      final CachePolicies<Key, CompletableFuture<? extends DataRetrievalResponse>, ? extends CacheEntry<CompletableFuture<? extends DataRetrievalResponse>>> cachePolicies ) 
+	{
+		super( fedAccMan, cacheCapacity, cachePolicies );
+		cardinalityCache = new CardinalityCacheMap<>();
+	}
+
+	public FederationAccessManagerWithPersistedDiskCache( final FederationAccessManager fedAccMan,
+	                                                      final int cacheCapacity ) 
+	{
+		this( fedAccMan, cacheCapacity, new MyDefaultCachePolicies() );
+	}
+
+	/**
+	 * Creates a {@link FederationAccessManagerWithPersistedDiskCache} with a default configuration.
+	 */
+	public FederationAccessManagerWithPersistedDiskCache( final ExecutorService execService ) 
+	{
+		this( new AsyncFederationAccessManagerImpl( execService ), 100, new MyDefaultCachePolicies() );
+	}
+	
+
+	@Override
+	public CompletableFuture<CardinalityResponse> issueCardinalityRequest( final SPARQLRequest req,
+	                                                                       final SPARQLEndpoint fm )
```

Comment:
Shouldn't the other `issueCardinalityRequest` methods have an implementation in this class as well?

Reply:
Yes, they should. I'll add them and the corresponding tests.

Reply:
Is this ready for me to look at again?

Reply:
Not yet, but I will let you know when it's ready. Working on it now.

Reply:
Okay, then simply request another review from me when you are done.

Reply:
Fixed. See commit c7f0dca9cde939ee68d88d36eb8def601793bd2d

Reply:
@hartig This is now ready for another review.